libs/core/langchain_core/indexes/base.py

```diff
@@ -0,0 +1,369 @@
+                            succeeded.append(id_)
+                    return DeleteResponse(succeeded=succeeded, failed=failed)
+
+                def lazy_get_by_ids(self, ids: Iterable[str]) -> Iterable[Document]:
+                    for id in ids:
+                        yield self.store[id]
+
+                def yield_keys(
+                    self, *, prefix: Optional[str] = None
+                ) -> Union[Iterator[str]]:
+                    prefix = prefix or ""
+                    for key in self.store:
+                        if key.startswith(prefix):
+                            yield key
+    """  # noqa: E501
+
+    @abstractmethod
+    def upsert(
+        self,
+        # TODO: Iterable or Iterator?
+        documents: Iterable[Document],
+        *,
+        ids: Optional[Iterable[str]] = None,
+        **kwargs: Any,
+    ) -> UpsertResponse:
+        """Upsert documents to index."""
+
+    # FOR CONTRIBUTORS: Overwrite this in Index child classes that support native async upsertions.
+    async def aupsert(
+        self,
+        documents: AsyncIterable[Document],
+        *,
+        ids: Optional[AsyncIterable[str]] = None,
+        **kwargs: Any,
+    ) -> UpsertResponse:
+        """Upsert documents to index. Default implementation, runs sync upsert() in async executor."""
+        # TODO: how to convert AsyncIterable -> Iterable
+        return await run_in_executor(None, self.upsert, documents, ids=ids, **kwargs)
+
+    @abstractmethod
+    def delete_by_ids(self, ids: Iterable[str]) -> DeleteResponse:
+        """Delete documents by id.
+
+        Args:
+            ids: IDs of the documents to delete.
+
+        Returns:
+           A dict ``{"succeeded": [...], "failed": [...]}`` with the IDs of the
+           documents that were successfully deleted and the ones that failed to be
+           deleted.
+        """
+
+    # FOR CONTRIBUTORS: Overwrite this in Index child classes that support native async deletions.
+    async def adelete_by_ids(self, ids: AsyncIterable[str]) -> DeleteResponse:
+        """Upsert documents to index. Default implementation, runs sync delete_by_ids() in async executor."""
+        return await run_in_executor(None, self.delete_by_ids, ids)
+
+    @abstractmethod
+    def lazy_get_by_ids(self, ids: Iterable[str]) -> Iterable[Document]:
+        """Lazily get documents by id.
+
+        Args:
+            ids: IDs of the documents to get.
+
+        Yields:
+           Document
+        """
+
+    # FOR CONTRIBUTORS: Overwrite this in Index child classes that support native async get.
+    async def alazy_get_by_ids(
+        self, ids: AsyncIterable[str]
+    ) -> AsyncIterable[Document]:
+        """Lazily get documents by id.
+
+        Default implementation, runs sync () in async executor.
+
+        Args:
+            ids: IDs of the documents to get.
+
+        Yields:
+           Document
+        """
+        return await run_in_executor(None, self.lazy_get_by_ids, ids)
```

Comment:
run_in_executor returns an Iterable, not an AsyncIterable.
I think run_in_executor(None, iterator, next) should be called in a loop